mithril-client/src/certificate_client.rs

```diff
@@ -0,0 +1,513 @@
+    },
+    entities::Certificate,
+    messages::CertificateMessage,
+};
+
+#[cfg(test)]
+use mockall::automock;
+
+/// Aggregator client for the Certificate
+pub struct CertificateClient {
+    aggregator_client: Arc<dyn AggregatorClient>,
+    retriever: Arc<InternalCertificateRetriever>,
+    verifier: Arc<dyn CertificateVerifier>,
+}
+
+/// API that defines how to validate certificates.
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait CertificateVerifier: Sync + Send {
+    /// Validate the chain starting with the given certificate.
+    async fn verify_chain(&self, certificate: &MithrilCertificate) -> MithrilResult<()>;
+}
+
+impl CertificateClient {
+    /// Constructs a new `CertificateClient`.
+    pub fn new(
+        aggregator_client: Arc<dyn AggregatorClient>,
+        verifier: Arc<dyn CertificateVerifier>,
+        logger: Logger,
+    ) -> Self {
+        let retriever = Arc::new(InternalCertificateRetriever {
+            aggregator_client: aggregator_client.clone(),
+            logger,
+        });
+
+        Self {
+            aggregator_client,
+            retriever,
+            verifier,
+        }
+    }
+
+    /// Fetch a list of signed certificates
+    pub async fn list(&self) -> MithrilResult<Vec<MithrilCertificateListItem>> {
+        let response = self
+            .aggregator_client
+            .get_content(AggregatorRequest::ListCertificates)
+            .await
+            .with_context(|| "CertificateClient can not get the certificate list")?;
+        let items = serde_json::from_str::<Vec<MithrilCertificateListItem>>(&response)
+            .with_context(|| "CertificateClient can not deserialize certificate list")?;
+
+        Ok(items)
+    }
+
+    /// Get a single certificate full information from the aggregator.
+    pub async fn get(&self, certificate_hash: &str) -> MithrilResult<Option<MithrilCertificate>> {
+        self.retriever.get(certificate_hash).await
+    }
+
+    /// Validate the chain starting with the certificate with given `certificate_hash`, return the certificate if
+    /// chain valid.
+    ///
+    /// This method will fails if no certicate exists with the given `certificate_hash`.
+    pub async fn verify_chain(&self, certificate_hash: &str) -> MithrilResult<MithrilCertificate> {
+        let certificate = self.retriever.get(certificate_hash).await?.ok_or(anyhow!(
+            "No certificate exist for hash '{certificate_hash}'"
+        ))?;
+
+        self.verifier
+            .verify_chain(&certificate)
+            .await
+            .with_context(|| {
+                format!("Certicate chain of certificate '{certificate_hash}' is invalid")
+            })?;
+
+        Ok(certificate)
+    }
+}
+
+/// Internal type to implement the [InternalCertificateRetriever] trait and avoid a circular
+/// dependency between the [CertificateClient] and the [MithrilCertificateVerifier] that need
+/// a [CertificateRetriever] as a dependency.
```

Comment:
Update this comment to refer to locally aliased common types (CommonCertificateVerifier, ...)